src/ripple/ripple.js

```diff
@@ -109,25 +110,52 @@
       }
       this.setRippleXY(x, y);
       this.setRippleStyles(true);
-      requestAnimationFrame(this.animFrameHandler.bind(this));
+
+      if (this.timeout) {
+        window.clearTimeout(this.timeout);
+      }
+
+      // We've just clicked, clear past upHandler_ call record
+      this.rippleElement_.mouseUp = false;
+      // Indicate to upHandler_ that ripple hasn't run for this.Constant_.DURATION
+      this.rippleElement_.animationCompleted = false;
+
+      window.requestAnimationFrame(function() {
+        this.animFrameHandler();
+
+        // Allow the animation to complete, even for tap events, which may
+        // trigger multiple mouseup with final event.detail == 1
+        // Ex: mac trackpad tap triggers 2 consecutive event.detail == 0 events
+        // followed by an event.detail=1 mouseup event
+        // we set the animation timeout inside the frame request to
+        // prevent animation flashing during long ui-blocking operations
+        this.timeout = window.setTimeout(function() {
+          // Check if use holding click event (or touchend not yet called)
+          if (this.rippleElement_.mouseUp) {
+            window.requestAnimationFrame(function() {
+              this.rippleElement_.classList.remove(this.CssClasses_.IS_VISIBLE);
+            }.bind(this));
+          }
+          // Allow upHandler_ to clear animation
+          this.rippleElement_.animationCompleted = true;
+        }.bind(this), Number(this.Constant_.ANIMATION_LENGTH));
+      }.bind(this));
     }
   };
 
   /**
    * Handle mouse / finger up on element.
    *
-   * @param {Event} event The event that fired.
    * @private
    */
-  MaterialRipple.prototype.upHandler_ = function(event) {
-    // Don't fire for the artificial "mouseup" generated by a double-click.
-    if (event && event.detail !== 2) {
-      // Allow a repaint to occur before removing this class, so the animation
-      // shows for tap events, which seem to trigger a mouseup too soon after
-      // mousedown.
-      window.setTimeout(function() {
-        this.rippleElement_.classList.remove(this.CssClasses_.IS_VISIBLE);
-      }.bind(this), 0);
+  MaterialRipple.prototype.upHandler_ = function() {
+    // set mouseup to allow downhanlder to remove visible state
```

Comment:
Please capitalize properly, and terminate with a period: `Set mouseUp to allow downHanlder to remove visible state.`